src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -271,18 +319,76 @@ public void CopyChangedValues(TStructuralType original)
                 throw Error.ArgumentNull("original");
             }
 
+            // Delta parameter type cannot be derived type of original
+            // to prevent unrecognizable information from being applied to original resource.
             if (!_entityType.IsAssignableFrom(original.GetType()))
             {
                 throw Error.Argument("original", SRResources.DeltaTypeMismatch, _entityType, original.GetType());
             }
 
-            PropertyAccessor<TStructuralType>[] propertiesToCopy = GetChangedPropertyNames().Select(s => _allProperties[s]).ToArray();
+            RuntimeHelpers.EnsureSufficientExecutionStack();
+
+            // For regular non-structural properties at current level.
+            PropertyAccessor<TStructuralType>[] propertiesToCopy =
+                this._changedProperties.Select(s => _allProperties[s]).ToArray();
             foreach (PropertyAccessor<TStructuralType> propertyToCopy in propertiesToCopy)
             {
                 propertyToCopy.Copy(_instance, original);
             }
 
             CopyChangedDynamicValues(original);
+
+            // For nested resources.
+            foreach (string nestedResourceName in _deltaNestedResources.Keys)
+            {
+                // Patch for each nested resource changed under this TStructuralType.
+                dynamic deltaNestedResource = _deltaNestedResources[nestedResourceName];
+                dynamic originalNestedResource = null;
+                if (!TryGetPropertyRef(original, nestedResourceName, out originalNestedResource))
+                {
+                    throw Error.Argument(nestedResourceName, "Cannot find nested resource name '{0}' in parent resource type '{1}'",
+                        nestedResourceName, original.GetType());
+                }
+
+                if (originalNestedResource == null)
+                {
+                    // When patching original target of null value, directly set nested resource.
+                    dynamic deltaObject = _deltaNestedResources[nestedResourceName];
+                    dynamic instance = deltaObject.GetInstance();
+
+                    // Recursively patch up the instance with the nested resources.
+                    deltaObject.CopyChangedValues(instance);
+
+                    _allProperties[nestedResourceName].SetValue(original, instance);
+                }
+                else
+                {
+                    // Recursively patch the subtree.
+                    deltaNestedResource.CopyChangedValues(originalNestedResource);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the property object by the specified name.
```

Comment:
>Gets the property object by the specified name. [](start = 12, length = 47)

resentence the comments because the method returns a bool?

Reply:
"Attempts to get"

---
In reply to: [190750618](https://github.com/OData/WebApi/pull/1445#discussion_r190750618) [](ancestors = 190750618)